File: typescript/src/yaml_codec.ts
/**
 * The YAML codec: bytes to a structure, and a structure to canonical bytes.
 *
 * THE LIBRARY EMITS AND THIS ADDS FOUR ADAPTERS. Nothing here writes a
 * character of YAML: layout, indentation, escaping and line breaks are all
 * js-yaml's, and its escape table is the one that ships. The adapters exist
 * because they preserve MEANING rather than layout, which is what
 * `packs-agree-on-structure-not-on-bytes` holds every pack to:
 *
 *   sort the keys                two runs over one structure must agree
 *   quote every string and key   `no`, `1.20`, `null` and `10` stay what they
 *                                were. The key is the sharp one: to a YAML 1.1
 *                                reader an unquoted `10:` is an integer key
 *   positional floats            FR-4.8, and `1e+20` read by a naive numeric
 *                                pattern yields 1
 *   null written as the word     an empty value and a missing one should not
 *                                look the same to a reader
 *
 * WHICH LIBRARY, AND WHY IT IS NOT THE OTHER ONE. Both maintained TypeScript
 * YAML libraries survive the acceptance check, which is a round trip of the
 * control-character fixture through the library's own reader. `yaml` 2.9.0 then
 * fails the requirement the check is a proxy for: its double-quoted escaping is
 * driven by `JSON.stringify`, which escapes nothing above U+001F, so U+007F,
 * U+0085, U+00A0, U+2028, U+2029 and U+FEFF are written RAW. Its own reader
 * takes them back, and Python's does not: ruamel refuses the whole document
 * with `unacceptable character #x007f`. A file no sibling pack can read is not
 * structural parity, so the round trip through one library is necessary and not
 * sufficient. Measured 2026-09-08.
 *
 * js-yaml escapes them as libyaml does, `\x7F`, `\N`, `\_`, `\L`, `\P` and
 * `U+FEFF`, and leaves U+200B alone, which is a printable character and is what
 * every other pack does with it.
 *
 * `forceQuotes` IS NOT THE QUOTING ADAPTER, and looks like it. It skips keys by
 * construction, and it quotes numbers and booleans as well as strings, which
 * makes the emitter write `!!int '3'` to keep the type. The hooks that do the
 * job are the schema's tag list and `transform`, which hands the emitter's own
 * AST over before it is rendered.
 */

import {
  CORE_SCHEMA,
  dump,
  floatCoreTag,
  intCoreTag,
  load,
  mapTag,
  nullCoreTag,
  realMapTag,
  SCALAR_STYLE,
  Schema,
  visit,
} from "js-yaml";

import { EncodeError, ParseError } from "./errors.ts";
import { canonicalNumberText } from "./float_text.ts";
import { asText, type Codec, type WrenchValue } from "./seams.ts";

const STR_TAG = "tag:yaml.org,2002:str";

/**
 * Which numbers are written without a decimal point.
 *
 * The same predicate decides the spelling and selects the tag, and that is what
 * keeps the emitter from writing an explicit `!!int`: a value the spelling
 * gives a `.0` must be selected by the float tag, or the text and the tag
 * disagree and the library says so in the output.
 *
 * JavaScript has one number type, so this is a property of the value rather
 * than of a type it was decoded as. Past 2^53 it is also the widening FR-4.10
 * asks for: a literal that arrived already rounded is written with a `.0`
 * rather than as an exact-looking integer it is not.
 */
function isPlainInteger(value: unknown): boolean {
  return typeof value === "number" && Number.isSafeInteger(value) &&
    !Object.is(value, -0);
}

// Adapters three and four. The CORE tags with `represent` replaced, so the
// reading side stays the library's except where the contract differs from it.
// deno-lint-ignore no-explicit-any
const INT_OUT: any = {
  ...intCoreTag,
  identify: isPlainInteger,
  represent: canonicalNumberText,
  // FR-4.11's YAML half. `Number("-0")` is a negative zero and an integer zero
  // cannot carry the sign, so the reader's answer is normalised here. Only the
  // float tag keeps a sign, and only for a source with a decimal point.
  resolve: (source: string, isExplicit: boolean, tagName: string) => {
    const value = intCoreTag.resolve(source, isExplicit, tagName);
    return Object.is(value, -0) ? 0 : value;
  },
};

/**
 * FR-4.11: in YAML `-0` is the integer zero and only `-0.0` carries a sign,
 * because YAML's type repository gives `0|-?[1-9][0-9]*` as the canonical
 * integer and zero has no signed variant there. The int tag already answers the
 * first half. js-yaml normalises the second to a positive zero, so this puts
 * the sign back, which is what makes a negative zero survive the round trip.
 */
// deno-lint-ignore no-explicit-any
const FLOAT_OUT: any = {
  ...floatCoreTag,
  identify: (data: unknown) => typeof data === "number",
  represent: canonicalNumberText,
  resolve: (source: string, isExplicit: boolean, tagName: string) => {
    const value = floatCoreTag.resolve(source, isExplicit, tagName);
    const signed = source.trimStart().startsWith("-") && /[.eE]/.test(source);
    return value === 0 && signed ? -0 : value;
  },
};

// deno-lint-ignore no-explicit-any
const NULL_OUT: any = { ...nullCoreTag, represent: () => "null" };

// deno-lint-ignore no-explicit-any
function replaced(tags: readonly any[]): any[] {
  return tags.map((tag) => {
    if (tag.tagName === intCoreTag.tagName) return INT_OUT;
    if (tag.tagName === floatCoreTag.tagName) return FLOAT_OUT;
    if (tag.tagName === nullCoreTag.tagName) return NULL_OUT;
    return tag;
  });
}

/** What the emitter is given. */
const WRITE_SCHEMA = new Schema(replaced(CORE_SCHEMA.tags));

/**
 * What the reader is given, and it differs in one tag.
 *
 * `realMapTag` builds a native `Map`, so a mapping key keeps the type it was
 * written with. The object-based tag turns `1:` into the key `"1"`, and a
 * coercion that is not reversible has to be refused rather than made (FR-2.9).
 * There is nowhere else to see it: by the time a mapping is a JavaScript object
 * the evidence is gone.
 */
const READ_SCHEMA = new Schema(
  replaced(CORE_SCHEMA.tags).map((
    tag,
  ) => (tag.tagName === mapTag.tagName ? realMapTag : tag)),
);

/**
 * Adapters one and two, through the emitter's own AST.
 *
 * A scalar carries its resolved tag, so the string ones are told to print
 * double-quoted and a number is left alone by construction rather than by
 * inspecting its text.
 *
 * The keys are sorted by their spelling, which is the same string comparison
 * every pack makes and is what puts `10` before `2`.
 */
// deno-lint-ignore no-explicit-any
function adapt(documents: any[]): void {
  // deno-lint-ignore no-explicit-any
  visit(documents, (node: any) => {
    if (node.kind === "scalar" && node.tag === STR_TAG) {
      node.style = SCALAR_STYLE.DOUBLE_QUOTED;
    }
    if (node.kind !== "mapping") return;
    // deno-lint-ignore no-explicit-any
    node.items.sort((one: any, two: any) => {
      const first = one.key.kind === "scalar" ? one.key.value : "";
      const second = two.key.kind === "scalar" ? two.key.value : "";
      return first < second ? -1 : first > second ? 1 : 0;
    });
  });
}

/**
 * `noRefs` is load-bearing. Left off, a structure holding the same object twice
 * is emitted with an anchor and an alias, which is a file this codec's own
 * reader then refuses. `flowLevel: -1` keeps block style at every depth
 * (FR-4.4), and `lineWidth: -1` turns folding off, because a folded scalar is a
 * line break inserted into a value.
 */
const WRITE_OPTIONS = {
  schema: WRITE_SCHEMA,
  flowLevel: -1,
  lineWidth: -1,
  noCompatMode: false,
  noRefs: true,
  skipInvalid: false,
  transform: adapt,
};

/**
 * `maxAliases: 0` refuses aliases rather than expanding them: a document that
 * expands to something larger than itself is a denial of service in a library
 * whose whole job is reading files from elsewhere. `json: false` keeps a
 * duplicate key an error rather than letting the last one win.
 */
const READ_OPTIONS = { schema: READ_SCHEMA, json: false, maxAliases: 0 };

export class YAMLCodec implements Codec {
  /**
   * Bytes into maps, lists and scalars.
   *
   * A `YAMLException` does not cross this boundary; the cause is kept.
   *
   * An empty document is `null` rather than a parse failure, which is what the
   * Python pack answers and what a caller reading a file somebody has emptied
   * expects. js-yaml calls it an error, so this is the one place the library's
   * reading is overridden rather than passed on.
   */
  decode(data: Uint8Array | string): WrenchValue {
    const text = asText(data);
    if (text.trim() === "") return null;
    try {
      return fromMaps(load(text, READ_OPTIONS));
    } catch (cause) {
      throw new ParseError(
        `could not parse the YAML: ${(cause as Error).message.split("\n")[0]}`,
        {
          cause,
        },
      );
    }
  }

  /** A structure into canonical bytes. */
  encode(value: WrenchValue): Uint8Array {
    try {
      refuseValuesWithNoCanonicalForm(value);
      return new TextEncoder().encode(dump(value, WRITE_OPTIONS));
    } catch (cause) {
      throw new EncodeError(
        `could not write canonical YAML: ${
          (cause as Error).message.split("\n")[0]
        }`,
        { cause },
      );
    }
  }
}

/**
 * The reader's `Map`s as plain objects, refusing a key that is not a string.
 *
 * An unquoted `10:` is a number to the reader and a quoted `"10":` is a string,
 * and the two are the same key by the time the document is a JavaScript object.
 * Refusing here is what stops a file meaning one thing on disk and another in
 * the structure a schema then validates.
 */
function fromMaps(value: unknown): WrenchValue {
  if (value instanceof Map) {
    const out: Record<string, WrenchValue> = {};
    for (const [key, item] of value) {
      if (typeof key !== "string") {
        throw new TypeError(`mapping key ${describe(key)} is not a string`);
      }
      out[key] = fromMaps(item);
    }
    return out;
  }
  if (Array.isArray(value)) return value.map(fromMaps);
  if (value === undefined) return null;
  return value as WrenchValue;
}

/** A refused key, named in a way a reader can find in the file. */
function describe(key: unknown): string {
  if (key === null) return "null";
  if (typeof key === "object") {
    return Array.isArray(key) ? "a sequence" : "a mapping";
  }
  return JSON.stringify(String(key));
}

/**
 * A value with no canonical form is refused rather than guessed at (FR-4.1).
 *
 * A JavaScript object can only have string keys, so what this reaches is the
 * value a caller could not write at all: an `undefined`, a function, a symbol,
 * a bigint, and a NaN or an infinity, which the emitter would otherwise spell
 * as `.nan` and `.inf` where JSON Schema has no way to represent either.
 */
function refuseValuesWithNoCanonicalForm(value: unknown): void {
  if (
    value === null || typeof value === "boolean" || typeof value === "string"
  ) return;
  if (typeof value === "number") {
    canonicalNumberText(value);
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) refuseValuesWithNoCanonicalForm(item);
    return;
  }
  if (typeof value === "object") {
    const record = value as Record<string, unknown>;
    for (const key of Object.keys(record)) {
      refuseValuesWithNoCanonicalForm(record[key]);
    }
    return;
  }
  throw new TypeError(`cannot write ${typeof value} in canonical form`);
}

/** The codec, named rather than inferred from a filename. */
export const YAML_CODEC: YAMLCodec = new YAMLCodec();

File: cpp/src/yaml_codec.cpp
// YAML, both directions, over libyaml's event interface.
//
// The emitter applies the four adapters and nothing else decides the layout:
// keys arrive sorted because `wrench::value` keeps them that way, every string
// and key is double quoted, a float is spelled by `canonical_float`, and null is
// written as the word. Two settings sit beside them, both settings and not code:
// the line width off, so a long scalar is never wrapped, and unicode passed
// through rather than escaped.
//
// The decoder does the half libyaml leaves to its caller. libyaml hands back a
// scalar's text and the flags saying how it was written, so turning `no`, `10`
// and `2026-01-01` into values is this file's work. The rule is one line:
// resolve a scalar only where it was written plain, so a quoted scalar and a
// tagged one are strings and a plain one is looked up in the table below.
//
//     no yes on off      strings         null ~ and empty    null
//     true false         booleans        10 010              integer 10
//     0x10               integer 16      1_000               integer 1000
//     1.20               float 1.2       1e3                 float 1000.0
//     2026-01-01         string          12:30               string
//     .inf .nan          floats, which canonical form then refuses on the way out
//
// Two shapes this does not reach, named so a reader stops looking. A merge key
// is an ordinary key spelled `<<`, so a document using one gets a mapping under
// that name rather than the merge YAML 1.1 describes. An explicit tag is not
// honoured beyond leaving the scalar as the string it was written as.

#include <yaml.h>

// Three of the includes below are inside jsoncons rather than at its surface.
// Building and reading a value instantiates them, so include-what-you-use names
// them, and the jig passes no mapping file that could say otherwise.
#include <algorithm>
#include <cctype>
#include <charconv>
#include <cstdint>
#include <cstring>
#include <expected>
#include <functional>
#include <jsoncons/basic_json.hpp>
#include <jsoncons/detail/make_obj_using_allocator.hpp>
#include <jsoncons/key_value.hpp>
#include <jsoncons/semantic_tag.hpp>
#include <jsoncons/utility/conversion.hpp>
#include <jsoncons/utility/read_number.hpp>
#include <limits>
#include <map>
#include <optional>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>

#include "wrench/canonical.hpp"
#include "wrench/codec.hpp"
#include "wrench/error.hpp"
#include "wrench/value.hpp"

namespace wrench {

namespace {

// ---- resolving a plain scalar ------------------------------------------------

// Which base a number's prefix declares.
constexpr int decimal = 10;
constexpr int octal = 8;
constexpr int hexadecimal = 16;

// The text with every underscore dropped, which is what a digit group separator
// amounts to once the value is read.
std::string without_separators(std::string_view text) {
    std::string kept(text);
    std::erase(kept, '_');
    return kept;
}

// One of the spellings the table gives for nothing at all.
bool is_null_word(std::string_view text) {
    return text.empty() || text == "null" || text == "Null" || text == "NULL" ||
           text == "~";
}

std::optional<bool> as_boolean(std::string_view text) {
    if (text == "true" || text == "True" || text == "TRUE") {
        return true;
    }
    if (text == "false" || text == "False" || text == "FALSE") {
        return false;
    }
    return std::nullopt;
}

// A number as it was written: the sign, the base its prefix declares, and the
// digits with the prefix and the separators gone.
struct digits {
    bool negative = false;
    int base = decimal;
    std::string body;
};

digits taken_apart(std::string_view text) {
    digits taken;
    std::string rest = without_separators(text);
    if (rest.starts_with('-') || rest.starts_with('+')) {
        taken.negative = rest.starts_with('-');
        rest.erase(0, 1);
    }
    if (rest.starts_with("0x") || rest.starts_with("0X")) {
        taken.base = hexadecimal;
        rest.erase(0, 2);
    } else if (rest.starts_with("0o") || rest.starts_with("0O")) {
        taken.base = octal;
        rest.erase(0, 2);
    }
    taken.body = std::move(rest);
    return taken;
}

// Every character is a digit the declared base has, and there is at least one.
bool all_digits(const digits& taken) {
    const int base = taken.base;
    return !taken.body.empty() &&
           std::ranges::all_of(taken.body, [base](const char one) {
               if (base == hexadecimal) {
                   return std::isxdigit(static_cast<unsigned char>(one)) != 0;
               }
               return one >= '0' && one <= (base == octal ? '7' : '9');
           });
}

// Past the signed 64-bit range the value widens to a float, and FR-4.8 then
// makes the widening visible in the spelling (FR-4.10). Base ten is handed to
// the standard library, which rounds the whole digit string correctly; the other
// two bases are accumulated, there being no such conversion for them.
double widened(const digits& taken) {
    double magnitude = 0;
    if (taken.base == decimal) {
        std::from_chars(
            taken.body.data(), taken.body.data() + taken.body.size(), magnitude);
    } else {
        for (const char one : taken.body) {
            const int digit =
                one <= '9' ? one - '0' : (std::tolower(one) - 'a') + decimal;
            magnitude = (magnitude * taken.base) + digit;
        }
    }
    return taken.negative ? -magnitude : magnitude;
}

std::optional<value> as_integer(std::string_view text) {
    const digits taken = taken_apart(text);
    if (!all_digits(taken)) {
        return std::nullopt;
    }
    std::uint64_t magnitude = 0;
    const std::from_chars_result read =
        std::from_chars(taken.body.data(),
                        taken.body.data() + taken.body.size(),
                        magnitude,
                        taken.base);
    const auto limit =
        static_cast<std::uint64_t>(std::numeric_limits<std::int64_t>::max());
    if (read.ec == std::errc{} && magnitude <= limit + (taken.negative ? 1U : 0U)) {
        const auto signed_magnitude = static_cast<std::int64_t>(magnitude);
        return value(taken.negative ? -signed_magnitude : signed_magnitude);
    }
    return value(widened(taken));
}

// How many digits stand at `at`, which is what the float shape is counted in.
std::size_t digit_run(std::string_view text, std::size_t at) {
    std::size_t seen = 0;
    while (at + seen < text.size() && text[at + seen] >= '0' &&
           text[at + seen] <= '9') {
        ++seen;
    }
    return seen;
}

// Whether the text is a decimal float and nothing else. The standard library's
// conversion accepts `inf` and `nan` as words, which YAML spells `.inf` and
// `.nan`, so the shape is checked here rather than left to it.
bool looks_like_float(std::string_view text) {
    std::size_t at = 0;
    if (text.starts_with('-') || text.starts_with('+')) {
        at = 1;
    }
    const std::size_t before = digit_run(text, at);
    at += before;
    std::size_t after = 0;
    bool dotted = false;
    if (at < text.size() && text[at] == '.') {
        dotted = true;
        after = digit_run(text, at + 1);
        at += 1 + after;
    }
    bool powered = false;
    if (at < text.size() && (text[at] == 'e' || text[at] == 'E')) {
        std::size_t power = at + 1;
        if (power < text.size() && (text[power] == '-' || text[power] == '+')) {
            ++power;
        }
        const std::size_t run = digit_run(text, power);
        powered = run > 0;
        at = power + run;
    }
    return at == text.size() && before + after > 0 && (dotted || powered);
}

std::optional<value> as_float(std::string_view text) {
    const std::string cleaned = without_separators(text);
    std::string_view body = cleaned;
    const bool negative = body.starts_with('-');
    if (negative || body.starts_with('+')) {
        body.remove_prefix(1);
    }
    if (body == ".inf" || body == ".Inf" || body == ".INF") {
        const double infinity = std::numeric_limits<double>::infinity();
        return value(negative ? -infinity : infinity);
    }
    if (body == ".nan" || body == ".NaN" || body == ".NAN") {
        return value(std::numeric_limits<double>::quiet_NaN());
    }
    if (!looks_like_float(cleaned)) {
        return std::nullopt;
    }
    // The sign is applied here rather than left to the conversion, which takes a
    // leading minus and refuses a leading plus. Handing it the whole text read
    // `+1.5` as nothing at all.
    double number = 0;
    std::from_chars(body.data(), body.data() + body.size(), number);
    return value(negative ? -number : number);
}

// The table in the header comment, applied to a scalar written plain. Anything
// that is none of these is the string it was written as, which is what keeps a
// timestamp and a version number whole.
value resolved(std::string_view text) {
    if (is_null_word(text)) {
        return value::null();
    }
    if (const std::optional<bool> answer = as_boolean(text)) {
        return {*answer};
    }
    if (const std::optional<value> number = as_integer(text)) {
        return *number;
    }
    if (const std::optional<value> number = as_float(text)) {
        return *number;
    }
    return {text};
}

// ---- reading -----------------------------------------------------------------

std::string_view text_of(const yaml_char_t* bytes, std::size_t length) {
    return {reinterpret_cast<const char*>(bytes), length};
}

std::string_view name_of(const yaml_char_t* anchor) {
    if (anchor == nullptr) {
        return {};
    }
    return {reinterpret_cast<const char*>(anchor),
            std::strlen(reinterpret_cast<const char*>(anchor))};
}

// The parser, the event it is sitting on, and the whole of libyaml's lifetime
// rules in one place: every event is deleted before the next is read, and the
// parser is deleted whichever way the read goes.
class stream {
   public:
    explicit stream(std::string_view bytes) {
        yaml_parser_initialize(&parser_);
        yaml_parser_set_input_string(
            &parser_,
            reinterpret_cast<const unsigned char*>(bytes.data()),
            bytes.size());
    }

    ~stream() {
        yaml_event_delete(&event_);
        yaml_parser_delete(&parser_);
    }

    stream(const stream&) = delete;
    stream& operator=(const stream&) = delete;
    stream(stream&&) = delete;
    stream& operator=(stream&&) = delete;

    // The next event, or false where the document could not be read.
    bool next() {
        yaml_event_delete(&event_);
        event_ = {};
        return yaml_parser_parse(&parser_, &event_) != 0;
    }

    [[nodiscard]] const yaml_event_t& at() const { return event_; }

    // What libyaml said, with where it said it, which is the half of a parse
    // failure a person acts on.
    [[nodiscard]] failure problem() const {
        std::string message = parser_.problem != nullptr
                                  ? parser_.problem
                                  : "the document could not be read";
        message += " at line ";
        message += std::to_string(parser_.problem_mark.line + 1);
        message += " column ";
        message += std::to_string(parser_.problem_mark.column + 1);
        return {.step = step::parse, .message = std::move(message), .cause = {}};
    }

   private:
    yaml_parser_t parser_{};
    yaml_event_t event_{};
};

failure refused(std::string what) {
    return {.step = step::parse, .message = std::move(what), .cause = {}};
}

// Anchors seen so far, by the name each was declared under, so an alias is the
// value its anchor held.
using anchors = std::map<std::string, value, std::less<>>;

result<value> read_node(stream& in, anchors& seen);

result<value> read_sequence(stream& in, anchors& seen) {
    value list = value::make_array();
    while (true) {
        if (!in.next()) {
            return std::unexpected(in.problem());
        }
        if (in.at().type == YAML_SEQUENCE_END_EVENT) {
            return list;
        }
        result<value> element = read_node(in, seen);
        if (!element.has_value()) {
            return std::unexpected(element.error());
        }
        list.push_back(std::move(*element));
    }
}

// A key is a string or the document is refused (FR-2.9). Stringifying one is
// not reversible, so the pack does not do it.
result<std::string> read_key(stream& in, anchors& seen) {
    result<value> key = read_node(in, seen);
    if (!key.has_value()) {
        return std::unexpected(key.error());
    }
    if (!key->is_string()) {
        return std::unexpected(refused(
            "a mapping key that is not a string is refused: " + key->to_string()));
    }
    return key->as_string();
}

result<value> read_mapping(stream& in, anchors& seen) {
    value mapping = value();
    while (true) {
        if (!in.next()) {
            return std::unexpected(in.problem());
        }
        if (in.at().type == YAML_MAPPING_END_EVENT) {
            return mapping;
        }
        result<std::string> key = read_key(in, seen);
        if (!key.has_value()) {
            return std::unexpected(key.error());
        }
        if (!in.next()) {
            return std::unexpected(in.problem());
        }
        result<value> held = read_node(in, seen);
        if (!held.has_value()) {
            return std::unexpected(held.error());
        }
        mapping[*key] = std::move(*held);
    }
}

// The alias's value, or a refusal where nothing declared that anchor. libyaml
// reports an undefined alias itself, so this answers the case where a document
// aliases an anchor declared later.
result<value> read_alias(const stream& in, const anchors& seen) {
    const std::string_view anchor = name_of(in.at().data.alias.anchor);
    const auto found = seen.find(anchor);
    if (found == seen.end()) {
        return std::unexpected(
            refused("an alias names an anchor that is not declared yet: " +
                    std::string(anchor)));
    }
    return found->second;
}

// The node the stream is sitting on, whole, with the stream left on its last
// event. An anchor is recorded once the node it names is complete, which is what
// makes an alias to a mapping the mapping and not a fragment of it.
result<value> read_node(stream& in, anchors& seen) {
    const yaml_event_t& event = in.at();
    if (event.type == YAML_ALIAS_EVENT) {
        return read_alias(in, seen);
    }
    std::string anchor;
    result<value> held;
    if (event.type == YAML_SCALAR_EVENT) {
        anchor = name_of(event.data.scalar.anchor);
        const std::string_view text =
            text_of(event.data.scalar.value, event.data.scalar.length);
        held = event.data.scalar.plain_implicit != 0 ? resolved(text) : value(text);
    } else if (event.type == YAML_SEQUENCE_START_EVENT) {
        anchor = name_of(event.data.sequence_start.anchor);
        held = read_sequence(in, seen);
    } else if (event.type == YAML_MAPPING_START_EVENT) {
        anchor = name_of(event.data.mapping_start.anchor);
        held = read_mapping(in, seen);
    } else {
        return std::unexpected(
            refused("a document holds something that is not a value"));
    }
    if (held.has_value() && !anchor.empty()) {
        seen.insert_or_assign(anchor, *held);
    }
    return held;
}

// Past the document: what follows the one value a file holds.
//
// A stream may carry several documents and this codec reads one, because a call
// answering with a value has nowhere to put the second. The refusal says so
// rather than returning the first and dropping the rest.
result<void> read_to_end(stream& in) {
    if (!in.next()) {
        return std::unexpected(in.problem());
    }
    if (in.at().type != YAML_DOCUMENT_END_EVENT) {
        return std::unexpected(refused("a document holds more than one value"));
    }
    if (!in.next()) {
        return std::unexpected(in.problem());
    }
    if (in.at().type != YAML_STREAM_END_EVENT) {
        return std::unexpected(refused("a file holds more than one document"));
    }
    return {};
}

// ---- writing -----------------------------------------------------------------

int keep(void* data, unsigned char* buffer, std::size_t size) {
    auto* out = static_cast<std::string*>(data);
    out->append(reinterpret_cast<const char*>(buffer), size);
    return 1;
}

// The emitter, its settings, and the bytes it has produced so far.
class sink {
   public:
    explicit sink(std::string* out) {
        yaml_emitter_initialize(&emitter_);
        yaml_emitter_set_output(&emitter_, keep, out);
        // Unicode through rather than escaped, and no line width at all, so a
        // long scalar is one line however long it is.
        yaml_emitter_set_unicode(&emitter_, 1);
        yaml_emitter_set_width(&emitter_, -1);
    }

    ~sink() { yaml_emitter_delete(&emitter_); }

    sink(const sink&) = delete;
    sink& operator=(const sink&) = delete;
    sink(sink&&) = delete;
    sink& operator=(sink&&) = delete;

    // One event out. The event is initialised by the caller and is spent here
    // whether or not the emitter took it.
    bool emit(yaml_event_t& event) { return yaml_emitter_emit(&emitter_, &event) != 0; }

    [[nodiscard]] failure problem() const {
        std::string message = emitter_.problem != nullptr
                                  ? emitter_.problem
                                  : "the value could not be written";
        return failure{
            .step = step::encode, .message = std::move(message), .cause = {}};
    }

   private:
    yaml_emitter_t emitter_{};
};

// A scalar, quoted or bare. Quoted is every string and every key, which is the
// adapter that makes escaping possible at all: a single-quoted scalar has no
// escapes, so a control character in one would be written raw. Bare is every
// other kind, which is what keeps a boolean a boolean.
bool write_scalar(sink& out, std::string_view text, bool quoted) {
    yaml_event_t event;
    // A scalar carrying neither a tag nor an implicit flag is refused by
    // libyaml, and the flag that has to say the tag is implied is the one
    // matching the style.
    yaml_scalar_event_initialize(
        &event,
        nullptr,
        nullptr,
        reinterpret_cast<yaml_char_t*>(const_cast<char*>(text.data())),
        static_cast<int>(text.size()),
        quoted ? 0 : 1,
        quoted ? 1 : 0,
        quoted ? YAML_DOUBLE_QUOTED_SCALAR_STYLE : YAML_PLAIN_SCALAR_STYLE);
    return out.emit(event);
}

result<void> write_node(sink& out, const value& document);

result<void> write_sequence(sink& out, const value& document) {
    yaml_event_t event;
    yaml_sequence_start_event_initialize(
        &event, nullptr, nullptr, 1, YAML_BLOCK_SEQUENCE_STYLE);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    for (const value& element : document.array_range()) {
        const result<void> written = write_node(out, element);
        if (!written.has_value()) {
            return written;
        }
    }
    yaml_sequence_end_event_initialize(&event);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    return {};
}

result<void> write_mapping(sink& out, const value& document) {
    yaml_event_t event;
    yaml_mapping_start_event_initialize(
        &event, nullptr, nullptr, 1, YAML_BLOCK_MAPPING_STYLE);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    // The members arrive sorted by key, that being a property of the value type
    // rather than something this loop does, which is the first adapter.
    for (const auto& member : document.object_range()) {
        if (!write_scalar(out, member.key(), true)) {
            return std::unexpected(out.problem());
        }
        const result<void> written = write_node(out, member.value());
        if (!written.has_value()) {
            return written;
        }
    }
    yaml_mapping_end_event_initialize(&event);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    return {};
}

// A string is a string, and not one of the library's other kinds wearing a
// string's storage: a tagged string is a big number, a byte string or a date to
// the library, none of which is in the model FR-2.9 fixes. The noesc tag its
// parser sets on a string needing no unescaping means a plain string, as none
// does.
bool plainly_a_string(const value& document) {
    return document.tag() == jsoncons::semantic_tag::none ||
           document.tag() == jsoncons::semantic_tag::noesc;
}

// A scalar in the model, spelled the one way canonical form allows. A value with
// no canonical form is refused rather than guessed at (FR-4.1), which is what
// NaN and the infinities reach.
result<void> write_leaf(sink& out, const value& document) {
    std::string spelled;
    bool quoted = false;
    if (document.is_null()) {
        spelled = "null";
    } else if (document.is_bool()) {
        spelled = document.as_bool() ? "true" : "false";
    } else if (document.is_double()) {
        const result<std::string> number = canonical_float(document.as_double());
        if (!number.has_value()) {
            return std::unexpected(number.error());
        }
        spelled = *number;
    } else if (document.is_int64() || document.is_uint64()) {
        spelled = document.to_string();
    } else if (document.is_string() && plainly_a_string(document)) {
        spelled = document.as_string();
        quoted = true;
    } else {
        return std::unexpected(
            failure{.step = step::encode,
                    .message = "a value outside the model has no canonical form: " +
                               document.to_string(),
                    .cause = {}});
    }
    if (!write_scalar(out, spelled, quoted)) {
        return std::unexpected(out.problem());
    }
    return {};
}

result<void> write_node(sink& out, const value& document) {
    if (document.is_object()) {
        return write_mapping(out, document);
    }
    if (document.is_array()) {
        return write_sequence(out, document);
    }
    return write_leaf(out, document);
}

// The document, from stream start to stream end. Implicit at both ends, so
// nothing writes a `---` or a `...` a reader would have to ignore.
result<void> write_document(sink& out, const value& document) {
    yaml_event_t event;
    yaml_stream_start_event_initialize(&event, YAML_UTF8_ENCODING);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    yaml_document_start_event_initialize(&event, nullptr, nullptr, nullptr, 1);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    const result<void> written = write_node(out, document);
    if (!written.has_value()) {
        return written;
    }
    yaml_document_end_event_initialize(&event, 1);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    yaml_stream_end_event_initialize(&event);
    if (!out.emit(event)) {
        return std::unexpected(out.problem());
    }
    return {};
}

}  // namespace

result<value> yaml_codec::decode(std::string_view bytes) const {
    stream in(bytes);
    if (!in.next()) {
        return std::unexpected(in.problem());
    }
    if (!in.next()) {
        return std::unexpected(in.problem());
    }
    // A file holding nothing at all is a document holding nothing at all.
    if (in.at().type == YAML_STREAM_END_EVENT) {
        return value::null();
    }
    if (!in.next()) {
        return std::unexpected(in.problem());
    }
    anchors seen;
    result<value> document = read_node(in, seen);
    if (!document.has_value()) {
        return document;
    }
    const result<void> ended = read_to_end(in);
    if (!ended.has_value()) {
        return std::unexpected(ended.error());
    }
    return document;
}

result<std::string> yaml_codec::encode(const value& document) const {
    std::string out;
    sink into(&out);
    const result<void> written = write_document(into, document);
    if (!written.has_value()) {
        return std::unexpected(written.error());
    }
    return out;
}

const codec& yaml() {
    static const yaml_codec one;
    return one;
}

}  // namespace wrench
